File: AppAlmacen-Java/src/main/java/almacen/RaftNode.java
package almacen;

import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.io.*;
import java.util.concurrent.*;

public class RaftNode {
    private final String[] peers;
    private RaftRole role = RaftRole.FOLLOWER;
    private int currentTerm = 0;
    private String votedFor = null;
    private List<LogEntry> log = new ArrayList<>();

    private String leaderId = null;
    private final String ip = "http://127.0.0.1:";
    private int port = -1;

    // Scheduler para heartbeats
    private final ScheduledExecutorService heartbeatScheduler = Executors.newScheduledThreadPool(1);
    // Executor separado para replicación
    private final ExecutorService replicationExecutor = Executors.newFixedThreadPool(3);

    private long lastHeartbeat = System.currentTimeMillis();

    public RaftNode(int port, String[] peers) {
        this.peers = peers;
        this.port = port;
        startFollowerTimeout();
        System.out.println("RAFT: Nodo creado con "+peers.length+" peers ("+String.join(", ",peers)+")");
    }

    private void startFollowerTimeout() {
        heartbeatScheduler.scheduleAtFixedRate(() -> {
            long now = System.currentTimeMillis();
            System.out.println("RAFT: Role="+role+", LastHeartbeat="+(now-lastHeartbeat)+" Term="+currentTerm);
            if (role == RaftRole.FOLLOWER && (now - lastHeartbeat > 3000)) {
                startElection();
            }
        }, 1000, 1000, TimeUnit.MILLISECONDS);
    }

    private void startElection() {
        System.out.println("RAFT: Iniciando elección");
        role = RaftRole.CANDIDATE;
        currentTerm++;
        votedFor = "self";
        int votesGranted = 1;
        int majority = (peers.length / 2) + 1;

        boolean anyPeerAvailable = false;
        for (String peer : peers) {
            if (askForVote(peer)) {
                votesGranted++;
                anyPeerAvailable = true;
            }
        }
        if(!anyPeerAvailable) {
            System.out.println("RAFT: No hay otros nodos disponibles, convirtiéndonos en lider");
            becomeLeader();
        }
        else if (votesGranted >= majority) {
            System.out.println("RAFT: Elección ganada, votes="+votesGranted);
            becomeLeader();
        } else {
            role = RaftRole.FOLLOWER;
            System.out.println("RAFT: Elección perdida");
        }
    }

    private boolean askForVote(String peer) {
        try {
            System.out.println("RAFT: Pidiendo voto a "+peer);
            URL url = new URL(peer+"/requestVote");
            HttpURLConnection con = (HttpURLConnection) url.openConnection();
            con.setRequestMethod("POST");
            con.setDoOutput(true);
            con.setRequestProperty("Content-Type","application/json");

            String body = "{\"term\":"+currentTerm+", \"candidateId\":\"self\"}";
            try(OutputStream os = con.getOutputStream()){
                os.write(body.getBytes(StandardCharsets.UTF_8));
            }
            int code = con.getResponseCode();
            if (code == 200) {
                try (InputStream in = con.getInputStream()) {
                    String resp = new String(in.readAllBytes(), StandardCharsets.UTF_8);
                    return resp.contains("\"voteGranted\":true");
                }
            }
        } catch (Exception e) {
            // peer no disponible
        }
        return false;
    }

    private void becomeLeader() {
        role = RaftRole.LEADER;
        leaderId = this.ip + this.port;
        startHeartbeats();
    }

    private void startHeartbeats() {
        // El envío de heartbeats permanece en un hilo separado e independiente de la replicación
        heartbeatScheduler.scheduleAtFixedRate(() -> {
            if (role == RaftRole.LEADER) {
                sendHeartbeats();
            }
        }, 0, 1000, TimeUnit.MILLISECONDS);
    }

    private void sendHeartbeats() {
        for (String peer : peers) {
            sendAppendEntries(peer);
        }
    }

    private void sendAppendEntries(String peer) {
        try {
            URL url = new URL(peer + "/appendEntries");
            HttpURLConnection con = (HttpURLConnection) url.openConnection();
            con.setConnectTimeout(2000);
            con.setReadTimeout(2000);
            con.setRequestMethod("POST");
            con.setDoOutput(true);
            con.setRequestProperty("Content-Type", "application/json");
            String body = "{\"term\":" + currentTerm + ", \"leaderId\":\"" + leaderId + "\", \"entries\":[]}";
            try (OutputStream os = con.getOutputStream()) {
                os.write(body.getBytes(StandardCharsets.UTF_8));
            }
            con.getResponseCode();
        } catch (Exception e) {
            // Peer no disponible
        }
    }

    public synchronized String getStatus() {
        return "Role: "+role+", Term: "+currentTerm+", Leader: "+leaderId+", LogSize: "+log.size();
    }

    public synchronized void handleAppendEntriesRequest(int term, String leaderId) {
        if (term >= currentTerm) {
            currentTerm = term;
            this.leaderId = leaderId;
            role = RaftRole.FOLLOWER;
            lastHeartbeat = System.currentTimeMillis();
        }
    }

    public synchronized boolean handleRequestVote(int term, String candidateId) {
        if (term > currentTerm) {
            currentTerm = term;
            votedFor = null;
            role = RaftRole.FOLLOWER;
        }

        if (votedFor == null || votedFor.equals(candidateId)) {
            votedFor = candidateId;
            return true;
        }
        return false;
    }

    public synchronized int appendLogEntry(String command) {
        if (command == null) {
            System.err.println("appendLogEntry: 'command' es null, no se puede agregar al log");
            return -1;
        }
        LogEntry logEntry = new LogEntry(currentTerm, command);
        System.out.println("RAFT: Agregando entrada al log: " + logEntry);
        log.add(logEntry);
        int index = log.size() - 1;
        return index;
    }

    public void replicatedLogEntry(int index) {
        // Llamamos en un hilo separado para no bloquear el hilo de heartbeats
        replicationExecutor.submit(() -> {
            synchronized (this) {
                System.out.println("RAFT: Replicando entrada "+index);
                if (role != RaftRole.LEADER) return;
                if (index < 0 || index >= log.size()) return;
                String command = log.get(index).getCommand();

                for (String peer : peers) {
                    System.out.println("RAFT: Replicando entrada "+index+" a "+peer);
                    System.out.println("command: "+command);
                    replicateEntry(peer, index, command);
                }
            }
        });
    }

    private void replicateEntry(String peer, int index, String command) {
        try {
            URL url = new URL(peer+"/replicateEntry");
            HttpURLConnection con = (HttpURLConnection) url.openConnection();
            con.setConnectTimeout(2000);
            con.setReadTimeout(2000);
            con.setRequestMethod("POST");
            con.setDoOutput(true);
            con.setRequestProperty("Content-Type","application/json");
            String body = "{\"term\":" + currentTerm + ", \"leaderId\":\"" + leaderId + "\", \"index\":" + index + ", \"entry\":\"" + command + "\"}";
            try(OutputStream os = con.getOutputStream()){
                os.write(body.getBytes(StandardCharsets.UTF_8));
            }
            con.getResponseCode();
        } catch (Exception e) {
            // Peer no disponible
        }
    }

    public int redirectToLeader(String command) {
    System.out.println("RAFT: Redirigiendo petición al líder, que es " + leaderId);
    if (role == RaftRole.LEADER) return -1;
    try {
        URL url = new URL(leaderId + "/appendLogEntry");
        HttpURLConnection con = (HttpURLConnection) url.openConnection();
        con.setRequestMethod("POST");
        con.setDoOutput(true);
        con.setRequestProperty("Content-Type", "application/json");
        String body = "{\"entry\":\"" + command + "\"}";
        try (OutputStream os = con.getOutputStream()) {
            os.write(body.getBytes(StandardCharsets.UTF_8));
        }
        int code = con.getResponseCode();
        if (code == 200) {
            try (InputStream in = con.getInputStream()) {
                String resp = new String(in.readAllBytes(), StandardCharsets.UTF_8);
                System.out.println("RAFT: Respuesta del líder: " + resp);
                return Integer.parseInt(resp);
            }
        } else {
            System.err.println("RAFT: Error redirigiendo al líder, código de respuesta: " + code);
        }
    } catch (Exception e) {
        System.err.println("RAFT: Excepción redirigiendo al líder: " + e.getMessage());
        e.printStackTrace();
    }
    return -1;
}


    public boolean isLeader() {
        return role == RaftRole.LEADER;
    }

    public String getLeader() {
        return leaderId;
    }
}

File: AppVentasPython/raft_node.py
import requests
import threading
import time
import json
from flask import Flask, request, jsonify

app = Flask(__name__)

class RaftRole:
    FOLLOWER = "FOLLOWER"
    CANDIDATE = "CANDIDATE"
    LEADER = "LEADER"

class RaftNode:
    def __init__(self, port, peers):
        self.port = port
        self.peers = peers  # Lista de URLs de los peers
        self.role = RaftRole.FOLLOWER
        self.currentTerm = 0
        self.votedFor = None
        self.log = []
        self.leaderId = None
        self.lastHeartbeat = time.time()
        self.lock = threading.RLock()

        self.election_timeout = 3.0  # Tiempo en segundos
        self.heartbeat_interval = 1.0  # Tiempo en segundos

        print(f"[{self._current_time()}] [Nodo {self.port}] Inicializado como FOLLOWER con peers: {self.peers}")
        self.start_follower_timeout()

    def _current_time(self):
        return time.strftime("%d/%b/%Y %H:%M:%S", time.localtime())

    def start_follower_timeout(self):
        def follower_check():
            while True:
                time.sleep(1)
                with self.lock:
                    print(f"[{self._current_time()}] [Nodo {self.port}] Verificando seguidor...")
                    if self.role == RaftRole.FOLLOWER and (time.time() - self.lastHeartbeat > self.election_timeout):
                        print(f"[{self._current_time()}] [Nodo {self.port}] Temporizador de seguidor expirado.")
                        self.start_election()

        t = threading.Thread(target=follower_check, daemon=True)
        t.start()

    def start_election(self):
        with self.lock:
            self.role = RaftRole.CANDIDATE
            self.currentTerm += 1
            self.votedFor = f"http://127.0.0.1:{self.port}"  # ID único
            majority = (len(self.peers) // 2) + 1
            print(f"[{self._current_time()}] [Nodo {self.port}] Iniciando elección para term {self.currentTerm}. Buscando mayoría: {majority}")

        votesReceived = 1  # Cuenta el voto propio
        anyPeerAvailable = False

        def request_vote_from_peer(peer):
            nonlocal votesReceived, anyPeerAvailable
            try:
                body = {"term": self.currentTerm, "candidateId": self.votedFor}
                r = requests.post(peer + "/requestVote", json=body, timeout=2)
                if r.status_code == 200:
                    resp = r.json()
                    if resp.get("voteGranted", False):
                        with self.lock:
                            votesReceived += 1
                            anyPeerAvailable = True
                            print(f"[{self._current_time()}] [Nodo {self.port}] Voto recibido de {peer}. Total votos: {votesReceived}")
            except Exception as e:
                print(f"[{self._current_time()}] [Nodo {self.port}] Error al solicitar voto de {peer}: {e}")

        threads = []
        for peer in self.peers:
            # Pasar correctamente el argumento 'peer' usando args
            t = threading.Thread(target=request_vote_from_peer, args=(peer,), daemon=True)
            t.start()
            threads.append(t)

        for t in threads:
            t.join()

        with self.lock:
            if not anyPeerAvailable:
                print(f"[{self._current_time()}] [Nodo {self.port}] No hay otros nodos disponibles. Convirtiéndose en líder por defecto.")
                self.become_leader()
            elif votesReceived >= majority:
                print(f"[{self._current_time()}] [Nodo {self.port}] Ganó la elección con {votesReceived} votos.")
                self.become_leader()
            else:
                print(f"[{self._current_time()}] [Nodo {self.port}] Perdió la elección con {votesReceived} votos.")
                self.role = RaftRole.FOLLOWER

    def become_leader(self):
        self.role = RaftRole.LEADER
        self.leaderId = f"http://127.0.0.1:{self.port}"
        self.lastHeartbeat = time.time()
        print(f"[{self._current_time()}] [Nodo {self.port}] Se ha convertido en LEADER para term {self.currentTerm}. Iniciando heartbeats.")
        self.start_heartbeats()

    def start_heartbeats(self):
        def send_heartbeats():
            while True:
                time.sleep(self.heartbeat_interval)
                with self.lock:
                    if self.role != RaftRole.LEADER:
                        break
                    self.send_heartbeats_to_peers()

        t = threading.Thread(target=send_heartbeats, daemon=True)
        t.start()

    def send_heartbeats_to_peers(self):
        for p in self.peers:
            try:
                body = {"term": self.currentTerm, "leaderId": self.leaderId, "entries": []}
                r = requests.post(p + "/appendEntries", json=body, timeout=1)
                if r.status_code == 200:
                    print(f"[{self._current_time()}] [Nodo {self.port}] Heartbeat enviado a {p}")
            except requests.exceptions.Timeout:
                print(f"[{self._current_time()}] [Nodo {self.port}] Heartbeat timeout con {p}")
            except Exception as e:
                print(f"[{self._current_time()}] [Nodo {self.port}] Error al enviar heartbeat a {p}: {e}")

    def handle_append_entries(self, term, leaderId):
        with self.lock:
            if term >= self.currentTerm:
                if term > self.currentTerm or self.leaderId != leaderId:
                    self.leaderId = leaderId
                    self.role = RaftRole.FOLLOWER
                    self.currentTerm = term
                    self.votedFor = None
                    print(f"[{self._current_time()}] [Nodo {self.port}] Recibió appendEntries de {leaderId} con term {term}. Convirtiéndose en FOLLOWER.")
                self.lastHeartbeat = time.time()
            else:
                print(f"[{self._current_time()}] [Nodo {self.port}] Recibió appendEntries con term {term} menor al actual {self.currentTerm}. Ignorando.")

    def handle_request_vote(self, term, candidateId):
        with self.lock:
            if term > self.currentTerm:
                print(f"[{self._current_time()}] [Nodo {self.port}] Term {term} recibido es mayor que el actual {self.currentTerm}. Actualizando term y convirtiéndose en FOLLOWER.")
                self.currentTerm = term
                self.votedFor = None
                self.role = RaftRole.FOLLOWER

            voteGranted = False
            if (self.votedFor is None or self.votedFor == candidateId) and term >= self.currentTerm:
                self.votedFor = candidateId
                voteGranted = True
                print(f"[{self._current_time()}] [Nodo {self.port}] Votando por {candidateId} en term {term}")
            else:
                print(f"[{self._current_time()}] [Nodo {self.port}] No vota por {candidateId} en term {term}. Votado por: {self.votedFor}")

            return {"voteGranted": voteGranted}

    def append_log_entry(self, command):
        with self.lock:
            entry = {"term": self.currentTerm, "command": command}
            self.log.append(entry)
            print(f"[{self._current_time()}] [Nodo {self.port}] Añadida entrada al log: {entry}")
            return len(self.log) - 1

    def replicate_log_entry(self, index, command):
        def replicate():
            with self.lock:
                if self.role != RaftRole.LEADER:
                    print(f"[{self._current_time()}] [Nodo {self.port}] No es líder. No puede replicar entradas.")
                    return
            for p in self.peers:
                try:
                    body = {
                        "term": self.currentTerm,
                        "leaderId": self.leaderId,
                        "index": index,
                        "entry": command
                    }
                    r = requests.post(p + "/replicateEntry", json=body, timeout=2)
                    if r.status_code == 200:
                        print(f"[{self._current_time()}] [Nodo {self.port}] Replicada entrada a {p}")
                except Exception as e:
                    print(f"[{self._current_time()}] [Nodo {self.port}] Error al replicar entrada a {p}: {e}")

        t = threading.Thread(target=replicate, daemon=True)
        t.start()

    def is_leader(self):
        with self.lock:
            return self.role == RaftRole.LEADER

    def get_leader(self):
        with self.lock:
            return self.leaderId

    def redirect_to_leader(self, command):
        leader = self.get_leader()
        if leader is None:
            print(f"[{self._current_time()}] [Nodo {self.port}] No hay líder actual para redirigir la solicitud.")
            return -1
        try:
            body = {"entry": command}
            r = requests.post(leader + "/appendLogEntry", json=body, timeout=2)
            if r.status_code == 200:
                print(f"[{self._current_time()}] [Nodo {self.port}] Redirigido a líder {leader} para añadir entrada.")
                return int(r.text)
            else:
                print(f"[{self._current_time()}] [Nodo {self.port}] Error al redirigir a líder {leader}: Status {r.status_code}")
                return -1
        except Exception as e:
            print(f"[{self._current_time()}] [Nodo {self.port}] Error al redirigir a líder {leader}: {e}")
            return -1

    def get_status(self):
        with self.lock:
            status = f"Role: {self.role}, Term: {self.currentTerm}, Leader: {self.leaderId}, LogSize: {len(self.log)}"
            print(f"[{self._current_time()}] [Nodo {self.port}] Estado actual: {status}")
            return status